Source/WebKit/NetworkProcess/Downloads/cocoa/DownloadCocoa.mm

```diff
@@ -122,6 +134,59 @@
 #endif
     }
 }
+
+void Download::setPlaceholderURL(NSURL *placeholderURL)
+{
+    if (!placeholderURL)
+        return;
+
+    BOOL usingSecurityScopedURL = [placeholderURL startAccessingSecurityScopedResource];
+
+    NSError *error = nil;
+    NSData *data = [placeholderURL bookmarkDataWithOptions:(1<<24) includingResourceValuesForKeys:nil relativeToURL:nil error:&error];
+    if (error)
+        RELEASE_LOG_ERROR(Network, "Failed to create bookmark data: %{public}@", error);
+
+    if (usingSecurityScopedURL)
+        [placeholderURL stopAccessingSecurityScopedResource];
+
+    CompletionHandler<void()> completionHandler = [weakThis = WeakPtr { *this }, this] {
+        if (!weakThis)
+            return;
+        // Start updating download progress when the client has received the placeholder URL.
+        // Otherwise, the placeholder might have been deleted by the time the client receives it.
+        startUpdatingProgress();
+    };
+
+    sendWithAsyncReply(Messages::DownloadProxy::DidReceivePlaceholderURL(placeholderURL, span(data)), WTFMove(completionHandler));
+}
+
+void Download::setFinalURL(NSURL *finalURL)
+{
+    if (!finalURL)
+        return;
+
+    BOOL usingSecurityScopedURL = [finalURL startAccessingSecurityScopedResource];
+
+    NSError *error = nil;
+    NSData *data = [finalURL bookmarkDataWithOptions:0 includingResourceValuesForKeys:nil relativeToURL:nil error:&error];
+    if (error)
+        RELEASE_LOG_ERROR(Network, "Failed to create bookmark data: %{public}@", error);
+
+    if (usingSecurityScopedURL)
+        [finalURL stopAccessingSecurityScopedResource];
+
+    send(Messages::DownloadProxy::DidReceiveFinalURL(finalURL, span(data)));
+
+    didFinish();
+}
+
+void Download::startUpdatingProgress() const
+{
+    auto startObservingSelector = @selector(startObserving);
+    if ([m_progress respondsToSelector:startObservingSelector])
+        [m_progress performSelector:startObservingSelector];
```

Comment:
We usually just call the selector directly after a respondsToSelector check:

if ([m_progress respondsToSelector:selector(startObserving)])
        [m_progress startObserving];